// @wagmi/core/providers/jsonRpc/dist/wagmi-core-providers-jsonRpc.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ethers = require('ethers');

function jsonRpcProvider(_ref) {
  let {
    pollingInterval,
    priority,
    rpc,
    stallTimeout,
    static: static_ = true,
    weight
  } = _ref;
  return function (chain) {
    const rpcConfig = rpc(chain);
    if (!rpcConfig || rpcConfig.http === '') return null;
    return {
      chain: { ...chain,
        rpcUrls: { ...chain.rpcUrls,
          default: rpcConfig.http
        }
      },
      provider: () => {
        const RpcProvider = static_ ? ethers.providers.StaticJsonRpcProvider : ethers.providers.JsonRpcProvider;
        const provider = new RpcProvider(rpcConfig.http, {
          ensAddress: chain.ensAddress,
          chainId: chain.id,
          name: chain.network
        });
        if (pollingInterval) provider.pollingInterval = pollingInterval;
        return Object.assign(provider, {
          priority,
          stallTimeout,
          weight
        });
      },
      ...(rpcConfig.webSocket && {
        webSocketProvider: () => new ethers.providers.WebSocketProvider(rpcConfig.webSocket, chain.id)
      })
    };
  };
}

exports.jsonRpcProvider = jsonRpcProvider;
